src/metatrain/experimental/nanopet/trainer.py

```diff
@@ -0,0 +1,481 @@
+            ]
+        else:
+            train_samplers = [None] * len(train_datasets)
+            val_samplers = [None] * len(val_datasets)
+
+        # Create dataloader for the training datasets:
+        train_dataloaders = []
+        for dataset, sampler in zip(train_datasets, train_samplers):
+            train_dataloaders.append(
+                DataLoader(
+                    dataset=dataset,
+                    batch_size=self.hypers["batch_size"],
+                    sampler=sampler,
+                    shuffle=(
+                        sampler is None
+                    ),  # the sampler takes care of this (if present)
+                    drop_last=(
+                        sampler is None
+                    ),  # the sampler takes care of this (if present)
+                    collate_fn=collate_fn,
+                )
+            )
+        train_dataloader = CombinedDataLoader(train_dataloaders, shuffle=True)
+
+        # Create dataloader for the validation datasets:
+        val_dataloaders = []
+        for dataset, sampler in zip(val_datasets, val_samplers):
+            val_dataloaders.append(
+                DataLoader(
+                    dataset=dataset,
+                    batch_size=self.hypers["batch_size"],
+                    sampler=sampler,
+                    shuffle=False,
+                    drop_last=False,
+                    collate_fn=collate_fn,
+                )
+            )
+        val_dataloader = CombinedDataLoader(val_dataloaders, shuffle=False)
+
+        # Extract all the possible outputs and their gradients:
+        train_targets = (model.module if is_distributed else model).dataset_info.targets
+        outputs_list = []
+        for target_name, target_info in train_targets.items():
+            outputs_list.append(target_name)
+            for gradient_name in target_info.gradients:
+                outputs_list.append(f"{target_name}_{gradient_name}_gradients")
+        # Create a loss weight dict:
+        loss_weights_dict = {}
+        for output_name in outputs_list:
+            loss_weights_dict[output_name] = (
+                self.hypers["loss"]["weights"][
+                    to_external_name(output_name, train_targets)
+                ]
+                if to_external_name(output_name, train_targets)
+                in self.hypers["loss"]["weights"]
+                else 1.0
+            )
+        loss_weights_dict_external = {
+            to_external_name(key, train_targets): value
+            for key, value in loss_weights_dict.items()
+        }
+        loss_hypers = copy.deepcopy(self.hypers["loss"])
+        loss_hypers["weights"] = loss_weights_dict
+        logging.info(f"Training with loss weights: {loss_weights_dict_external}")
+
+        # Create a loss function:
+        loss_fn = TensorMapDictLoss(
+            **loss_hypers,
+        )
+
+        # Create an optimizer:
+        optimizer = torch.optim.Adam(
```

Comment:
PET uses AdamW, shall we also use it here?

Reply:
Yes, but then PET always sets the weight decay to zero, making it equivalent to Adam, no?